src/asl.ts

```diff
@@ -0,0 +1,720 @@
+   */
+  MaxConcurrency?: number;
+}
+
+export interface ParallelTask extends BaseTask {
+  Type: "Parallel";
+  Iterator: States;
+}
+
+/**
+ * Amazon States Language (ASL) Generator.
+ */
+export class ASL {
+  static readonly ContextName = "Amazon States Language";
+
+  readonly kind = ASL.ContextName;
+
+  /**
+   * Maps a {@link Stmt} to a unique State name. This is generated by passing over
+   * the AST and assigning each state a unique name. These names are then used to
+   * resolve state transitions.
+   */
+  readonly stateTransitions: Map<Stmt, string>;
+  readonly returnTransitions: Map<FunctionDecl | FunctionExpr, string>;
+  readonly throwTransitions: Map<FunctionDecl | FunctionExpr, string>;
+
+  readonly definition: StateMachine<States>;
+
+  constructor(
+    readonly scope: Construct,
+    readonly role: aws_iam.IRole,
+    readonly decl: FunctionDecl
+  ) {
+    let i = 0;
+    const nextName = () => `State${(i += 1)}`;
+    this.stateTransitions = new Map<Stmt, string>(bind(decl.body));
+    this.returnTransitions = new Map();
+    this.throwTransitions = new Map();
+
+    const states = this.execute(this.decl.body);
+
+    const start = this.getStateName(this.decl.body.statements[0]);
+
+    this.definition = {
+      StartAt: start,
+      States: {
+        ...states,
+        Success: {
+          Type: "Succeed",
+        },
+        Throw: {
+          Type: "Fail",
+          Error: "TODO",
+        },
+      },
+    };
+
+    function bind(stmt: Stmt | undefined): [Stmt, string][] {
+      if (stmt === undefined) {
+        return [];
+      } else if (stmt.kind === "BlockStmt") {
+        return stmt.statements.reduce(
+          (arr: [Stmt, string][], stmt) => arr.concat(bind(stmt)),
+          []
+        );
+      } else if (stmt.kind === "IfStmt") {
+        return [[stmt, nextName()], ...bind(stmt.then), ...bind(stmt._else)];
+      } else if (stmt.kind === "ExprStmt") {
+        if (stmt.expr.kind === "FunctionExpr") {
+          [[stmt, nextName()], ...bind(stmt.expr.body)];
+        }
+      } else if (stmt.kind === "ForInStmt" || stmt.kind === "ForOfStmt") {
+        return [[stmt, nextName()], ...bind(stmt.body)];
+      }
+      return [[stmt, nextName()]];
+    }
+  }
+
+  public execute(stmt: Stmt): States {
+    if (stmt.kind === "BlockStmt") {
+      return stmt.statements.reduce((states: States, s, i) => {
+        if (i < stmt.statements.length - 1) {
+          return {
+            ...states,
+            ...this.execute(s),
+          };
+        } else {
+          return {
+            ...states,
+            ...this.execute(s),
+          };
+        }
+      }, {});
+    } else if (stmt.kind === "BreakStmt") {
+      return {
+        [this.getStateName(stmt)]: {
+          Type: "Pass",
+          Next: this.next(findLoop(stmt)),
+        },
+      };
```

Comment:
Yeah it can most likely be optimized. I was erring on the side of caution since it wasn't clear in my head how this whole thing worked. I think it's safe to optimize this away now.